### PR TITLE
Fix: Throw error when dim is provided non integer value in sum intrinsic #4305

### PR DESCRIPTION
**Pull request by shubhneetgrover**
Towards : #4017 
Fixes: #4305

### CODE REVIEW

File: src/libasr/pass/intrinsic_array_function_registry.h [hunk 1]
```diff
@@ -596,6 +596,11 @@ static inline ASR::asr_t* create_ArrIntrinsic(
             if (args[2]) {
                 dim = args[2];
             }
+        } else{
+            append_error(diag, "`dim` argument to `" + intrinsic_func_name + "` intrinsic at (1) must be INTEGER",
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_array_function_registry.h:600]:
```suggestion
            append_error(diag, "`dim` argument to `" + intrinsic_func_name + "` must be integer",
```

Comment (HarshitaKalani) [reply src/libasr/pass/intrinsic_array_function_registry.h:600]: I think this needs to be handled args[2] is non-integer (at line 612)

**Review by HarshitaKalani — COMMENTED** [review]
Thanks for the PR! 
Can you please add a test for this?

File: src/libasr/pass/intrinsic_array_function_registry.h (hunk 1 above)
Comment (shubhneetgrover) [reply src/libasr/pass/intrinsic_array_function_registry.h:600]: Thanks for your suggestion.

As per my understanding args[2] will be always be mask and not dim. 

I didn't want to deviate the code fix from the issue requirment and also to keep it succinct.

File: tests/errors/dim_float_03.f90 [hunk 2]
```diff
@@ -0,0 +1,3 @@
+program main
+	print *, sum([1, 2, 3], 1.1, 1)
```
Comment (Thirumalai-Shaktivel) [inline tests/errors/dim_float_03.f90:2]:
```suggestion
	print *, sum([1, 2, 3], 1, 1.1)
```

File: tests/errors/dim_float_04.f90 [hunk 3]
```diff
@@ -0,0 +1,3 @@
+program main
+	print *, sum([1, 2, 3], 1.1, 1.1)
```
Comment (Thirumalai-Shaktivel) [inline tests/errors/dim_float_04.f90:2]:
```suggestion
	print *, sum([1, 2, 3], .true., 1.1)
```

**Review by Thirumalai-Shaktivel — APPROVED** [review]
LGTM!

I left minor suggestions to also check other errors

**Comment by shubhneetgrover** [comment]
This PR fixes the #4305 bug.
The root cause was the errors not being raised in case of invalid arguments to dim or mask. Fixed that by adding if checks before procedding with the code flow.
Test cases have been added for this fix.

File: tests/reference/asr-dim_float_02-2681d16.stderr [hunk 4]
```diff
@@ -0,0 +1,5 @@
+semantic error: Invalid argument type for `dim` or `mask`
```
Comment (certik) [inline tests/reference/asr-dim_float_02-2681d16.stderr:1]: This error is already tested at `float_01`.

File: tests/reference/asr-dim_float_03-ba04334.stderr [hunk 5]
```diff
@@ -0,0 +1,5 @@
+semantic error: Invalid argument type for `dim` or `mask`
```
Comment (certik) [inline tests/reference/asr-dim_float_03-ba04334.stderr:1]: already tested by float_01

File: tests/reference/asr-dim_float_04-03dc0c1.stderr [hunk 6]
```diff
@@ -0,0 +1,5 @@
+semantic error: Invalid argument type for `dim` or `mask`
```
Comment (certik) [inline tests/reference/asr-dim_float_04-03dc0c1.stderr:1]: already tested by float_01

**Review by certik — COMMENTED** [review]
I think we only need one test per error message. I marked the duplicated tests. I think 3 tests should be enough.

**Review by HarshitaKalani — APPROVED** [review]
Looks good to me. Thanks for the PR!

**Review by certik — APPROVED** [review]
I think this looks great, thanks for the PR!

**Pull request merged**